### ISSUE TITLE
Pygame Clock inaccurate

### ISSUE THREAD

**Issue by illume** [pygame/pygame#304]
Originally reported by: **Don Polettone (Bitbucket: [DonPolettone](https://bitbucket.org/DonPolettone), GitHub: [DonPolettone](https://github.com/DonPolettone))**

----------------------------------------

Hi all,

this is something that's been making me mad for months, if not years, and I hope someone can finally fix this:

The clock seems not doing well in pygame; animation stutters / jitters. First I thought I was doing something wrong, but when I digged deeper and broke it down to a test as simple as can be, I recognised it must be an issue with pygame itself.

Try THIS out:

```
#!python

import sys
import pygame
pygame.init()

RES = (640, 480)
FPS = 60
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
BLUE = (0, 0, 255)

display = pygame.display.set_mode(RES, pygame.FULLSCREEN)
clock = pygame.time.Clock()

bg_img = pygame.Surface(RES)
bg_img.fill(WHITE)
for y in (229, 290):
    for x in range(40, RES[0], 40):
        bg_img.set_at((x, y), BLUE)

def motion_test():

    rect = pygame.rect.Rect((0, 240), (40, 40))

    while True:

        display.blit(bg_img, (0, 0))
        
        for e in pygame.event.get():
            if (e.type == pygame.QUIT or
                e.type == pygame.KEYDOWN and e.key == pygame.K_ESCAPE):
                pygame.quit()
                sys.exit()
        
        rect.x += 1
        if rect.x >= RES[0]:
            rect.x = 0
        
        pygame.draw.rect(display, BLUE, rect)
        
        pygame.display.flip()
        clock.tick(FPS)

if __name__ == "__main__":
    motion_test()


```
Every few seconds, the rect stutters a bit. In a pixelated game, this looks awful. Especially when scrolling over pixely tilemaps. I tried it out on both Mac and Win - it is a general issue.

Can someone help/test/agree/decline/fix this ??

----------------------------------------
- Bitbucket: https://bitbucket.org/pygame/pygame/issue/304

**Related Docs**: https://www.pygame.org/docs/ref/time.html#pygame.time.Clock

**Comment by MyreMylar** [comment]
I was tinkering with this today and observed the same thing as described above - it seems the only way to get completely smooth screen scrolling/movement in pygame 1.9 is to trick the system into enabling vsynch to control the framerate. All the other combinations I've tried using the pygame clock you get a slight hitching every second or so.

However, this hacky method is no longer available in pygame 2 as the:

`os.environ['SDL_VIDEODRIVER'] = 'directx'`

...line is no longer available in SDL2 (the only available videodriver option on windows is now 'windows' see here: https://wiki.libsdl.org/FAQUsingSDL).

So in pygame 2 we now have the worst of both worlds, the same, seemingly pygame clock caused hitching, and the hacky way to bypass it by enabling vsynch is no longer available.

**Comment by robertpfeiffer** [comment]
@MyreMylar Did you disable the clock and use vsync only?

**Comment by MyreMylar** [comment]
@robertpfeiffer To get smooth scrolling in pygame 1 on windows? 

Sort of, from memory you need to stop limiting the frame rate with clock.tick(60) or whatever and let the v sync take care of doing that instead - but you don't need to remove the clock entirely. You can still run clock.tick() with no FPS limit argument and calculate the current framerate.

**Comment by robertpfeiffer** [comment]
Oh, yeah. Clock.tick(60) should not wait at all when vsync is on and the vsync frame rate is 50 Hz. But then the game only runs at 50 Hz instead of 60. On the other hand, if the screen refresh rate is 50 Hz and you have clock.tick(30) in there somewhere, you might get additional stutter from clock.tick. I think modern game engines solve this by only locking the frame rate at 144fps, and parameterising the update step with the elapsed time, or by having the game logic that needs a fixed time step run at 30/60FPS in its own thread with separate threads for UI/input, game logic, rendering, and networking.

We could enable vsync again, but we'd need to keep it turned off by default for backwards compatibility.

**Comment by MyreMylar** [comment]
Yes that is the sensible way to go. Add a flag to set mode or something
like that that lets users turn v sync on and off with it turned off by
default.

That way it is consistent, functionality wise, with pygame 1, but with a
less confusing interface for turning v sync on or off.
